pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
 }
 
 static npy_int64 get_long_attr(PyObject *o, const char *attr) {
+    // NB we are implicitly assuming that o is a Timedelta or Timestamp, or NaT
+
     npy_int64 long_val;
     PyObject *value = PyObject_GetAttrString(o, attr);
     long_val =
         (PyLong_Check(value) ? PyLong_AsLongLong(value) : PyLong_AsLong(value));
+
     Py_DECREF(value);
+
+    if (long_val == NPY_MIN_INT64) {
+        // i.e. o is NaT
+        return long_val;
+    }
+
+    // ensure we are in nanoseconds, similar to Timestamp._as_creso or _as_unit
+    PyObject* reso = PyObject_GetAttrString(o, "_creso");
+    // if (!PyLong_Check(reso)) {
```

Comment:
```suggestion
    if (!PyLong_Check(reso))
      return -1;
```

Might as well set this up now. The caller should be checking for -1 and PyErr_Occurred() (can be a separate PR if not)